src/main/kotlin/com/atlassian/performance/tools/awsinfrastructure/api/network/access/LocalPublicIpv4Provider.kt

```diff
@@ -0,0 +1,61 @@
+package com.atlassian.performance.tools.awsinfrastructure.api.network.access
+
+import com.atlassian.performance.tools.awsinfrastructure.Ipv4Validator
+import com.atlassian.performance.tools.awsinfrastructure.VerboseUrlReader
+import java.net.URL
+import java.util.function.Supplier
+
+class LocalPublicIpv4Provider private constructor(
+    private val servicesToQuery: List<URL>,
+    private val urlReader: (serviceUrl: URL) -> String?,
+    private val ipValidator: (ip: String) -> Boolean
+) : () -> String, Supplier<String> {
+    object Defaults {
+        val servicesToQuery: List<URL> = listOf(
+            URL("https://checkip.amazonaws.com")
+        )
+        val urlReader: (serviceUrl: URL) -> String? = VerboseUrlReader()
+        val ipValidator: (ip: String) -> Boolean = Ipv4Validator()
+    }
+
+    constructor() : this(
+        servicesToQuery = Defaults.servicesToQuery,
+        urlReader = Defaults.urlReader,
+        ipValidator = Defaults.ipValidator
+    )
+
+    override fun invoke() = get()
+
+    override fun get() = servicesToQuery
+        .asSequence()
+        .mapNotNull { serviceUrl -> urlReader(serviceUrl)?.trim() }
+        .filter { it.isNotEmpty() }
+        .filter { ipValidator(it) }
+        .toSet()
+        .let { set ->
+            if (set.isEmpty()) {
+                throw IllegalStateException("Queried services didn't report any valid public IP for this machine")
```

Comment:
I wonder, should we fail or just warn and go on? As I understand if we continue, the client machine won't get access non-SSH access to JPT nodes, which in many (most?) cases won't be an issue.

Reply:
Similarly to the `IllegalStateException ` triggered by `set.size != 1` I didn't think enough about this behavior to have a strong opinion. My concern with making the mechanism less strict around failures is that a silentish error like this could hide a cause while still manifesting other symptoms which are not obvious to lead towards the lack of public access. I'm seeing someone in the future failing to access JVM debug port of Jira for long time not realising to that the answer to the problem is in a single WARN line of the log (especially that there may be more WARN log lines).

The `LocalPublicIpv4Provider` is supposed to be public API and the `accessRequester` in both Server and DC formulas is overridable, so it should be rather easy to change the services used by the mechanism to the ones that the consumer has access to currently. Alternatively the initial `AccessRequester` can be disabled completely by setting it with `NoopAccessRequester`.

Reply:
@mzyromski-atlassian, @dagguh, @mgrzaslewicz what are your thoughts about this? 

The mechanism is using a external service to check public IP address of the test executor (local machine). Current approach will throw `IllegalStateException` if this service is not working. This is not cought by anything, so the test would fail if the service would be down. 

I picked only 1 service, however this mechanism can be overwritten in the `...Formula`'s `Builder` and it's possible to set multiple services. In case 2 services would report different public IP addresses for our local machine this would also throw `IllegalStateException` resulting in test failure. If some of the services fail to report an address (e.g. they are down) only addresses reported by other services will be taken into account (or as mentioned earlier it would throw if no other services are up).

Do you think the behavior should be changed? What'd be the best logic here?

Reply:
One thing I can do i to eagerly get the IP during the test static code initialisation. It would make the test fail fast in case of external service problem. Should this be done?
```
val accessRequester: AccessRequester = ForIpAccessRequester(LocalPublicIpv4Provider().get().let { ip -> { ip }})
```
instead of 
```
val accessRequester: AccessRequester = ForIpAccessRequester(LocalPublicIpv4Provider())
```

Reply:
Given that the service is actually provided by AWS, I guess we can expect it to be quite stable. I agree that failing fast might be a better approach than being surprised later. The eager IP fetch sounds like a plan too.

Reply:
- [ ] Resolve the public IP eagerly
- [ ] Explain what can be done by the API consumer in the error messages

Reply:
Spending a lot time lately with JPT I'd be frustrated not being able to log into the db or Jira host after some failure.
How often does it fail?

Reply:
> @mzyromski-atlassian, @dagguh, @mgrzaslewicz what are your thoughts about this?
> 
> The mechanism is using a external service to check public IP address of the test executor (local machine). Current approach will throw `IllegalStateException` if this service is not working. This is not cought by anything, so the test would fail if the service would be down.
> 
> I picked only 1 service, however this mechanism can be overwritten in the `...Formula`'s `Builder` and it's possible to set multiple services. In case 2 services would report different public IP addresses for our local machine this would also throw `IllegalStateException` resulting in test failure. If some of the services fail to report an address (e.g. they are down) only addresses reported by other services will be taken into account (or as mentioned earlier it would throw if no other services are up).
> 
> Do you think the behavior should be changed? What'd be the best logic here?

Spending a lot time lately with JPT I'd be frustrated not being able to log into the db or Jira host after some failure.
How often does it fail?

Reply:
> How often does it fail?
I don't expect that to ever happen, however you never know... with this we are making the whole code dependant on an external service functionality. Sure it can be overwritten, but we could end up in a situation in the future that for "old" versions of aws-infrastructure we would always need to overwrite it, because the service we are using no longer exists. We could quickfix it though.

I believe I will pick some backup services for this.

- [ ] Add backup public ip services
- [ ] Ad unit tests for the public ip service fallback

Reply:
> backup public ip services

YAGNI. Add backup only when the default service is actually flaky.

Reply:
> Resolve the public IP eagerly

Don't do any work during initialization. Especially if the effect of the default work can be overridden via builder.

Reply:
> should we fail or just warn and go on? 

That's a decision outside of `LocalPublicIpv4Provider`.

Reply:
> YAGNI. Add backup only when the default service is actually flaky.

I disagree. It's part of the solution quality to not rely on a single service by default. I'm still adding additional services, because it's low cost.

> Don't do any work during initialization. Especially if the effect of the default work can be overridden via builder.

I agree. Will skip this.

> That's a decision outside of `LocalPublicIpv4Provider`.

I'd like to avoid creating API out of exceptions. I will still leave the code as it is around it, because if someone would want to have it differently it's possible to overwrite this dependency. I'm only adding more verbose explanation of what can be done in case of given exception happening.